discovery/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,220 @@
+func (c *ZooKeeper) MarkForMaintenance(service *discovery.ServiceDefinition) {
+	c.Deregister(service)
+}
+
+// SendHeartbeat refreshes the TTL of this associated zookeeper node
+func (c *ZooKeeper) SendHeartbeat(service *discovery.ServiceDefinition) {
+	if err := c.registerService(service); err != nil {
+		log.Warnf("Error registering service %s: %s", service.Name, err)
+	}
+}
+
+func (c *ZooKeeper) parentPath(service *discovery.ServiceDefinition) string {
+	return fmt.Sprintf("%s/%s", c.Prefix, service.Name)
+}
+
+func (c *ZooKeeper) getNodeKey(service *discovery.ServiceDefinition) string {
+	return fmt.Sprintf("%s/%s", c.parentPath(service), service.ID)
+}
+
+func (c *ZooKeeper) getAppKey(appName string) string {
+	return fmt.Sprintf("%s/%s", c.Prefix, appName)
+}
+
+var zookeeperUpstreams = make(map[string][]ServiceNode)
+
+// CheckForUpstreamChanges checks another zookeeper node for changes
+func (c *ZooKeeper) CheckForUpstreamChanges(backendName, backendTag string) bool {
+	// TODO: is there a way to filter by tag in zookeeper?
+	services, err := c.getServices(backendName)
+	if err != nil {
+		log.Errorf("Failed to query %v: %s", backendName, err)
+		return false
+	}
+	didChange := zookeeperCompareForChange(zookeeperUpstreams[backendName], services)
+	if didChange || len(services) == 0 {
+		// We don't want to cause an onChange event the first time we read-in
+		// but we do want to make sure we've written the key for this map
+		zookeeperUpstreams[backendName] = services
+	}
+	return didChange
+}
+
+func (c *ZooKeeper) getServices(appName string) ([]ServiceNode, error) {
+	var services []ServiceNode
+
+	key := c.getAppKey(appName)
+	children, _, error := c.Connection.Children(key)
+	if error != nil {
+		return services, error
+	}
+	for i := range children {
+		path := fmt.Sprintf("%s/%s", key, children[i])
+		data, _, error := c.Connection.Get(path)
+		if error != nil {
+			return services, error
+		}
+		srv, error := decodeZooKeeperNodeValue(data)
+		if error != nil {
+			return services, error
+		}
+		services = append(services, srv)
+	}
+	return services, nil
+}
+
+// Compare the two arrays to see if the address or port has changed
+// or if we've added or removed entries.
+func zookeeperCompareForChange(existing, new []ServiceNode) (changed bool) {
+	if len(existing) != len(new) {
+		return true
+	}
+
+	sort.Sort(ByZooKeeperServiceID(existing))
+	sort.Sort(ByZooKeeperServiceID(new))
+	for i, ex := range existing {
+		if ex.Address != new[i].Address ||
+			ex.Port != new[i].Port {
+			return true
+		}
+	}
+	return false
+}
+
+func (c ZooKeeper) createParentPath(path string) error {
+	pathElements := strings.Split(path, "/")[1:]
+	sep := "/"
+	newPath := ""
+	for i := range pathElements {
+		newPath = strings.Join([]string{newPath, sep, pathElements[i]}, "")
+		if exists, _, _ := c.Connection.Exists(newPath); !exists {
+			if _, err := c.Connection.Create(newPath, nil, 0, zk.WorldACL(zk.PermAll)); err != nil {
```

Comment:
As far as I can tell, there’s no transaction support in ZK out of the box. @justenwalker do you have any ref handy?